dev/source/docs/ardupilot-on-linux-porting.rst

```diff
@@ -0,0 +1,142 @@
+
+An error message should appear on the ground station HUD complaining, “failed to init barometer”.
+
+Step 5 - peripheral functions and drivers
+=========================================
+
+Many boards will use the Linux userspace APIs for I2C, UART, and SPI and only need configuration.
+See your CPU and board's documentation for pin muxing and device tree configuration. The following peripherals tend to be more complicated.
+
+SPI Support
+-----------
+To configure SPI-based devices, the Linux userspace SPI APIs are used, needing only the per-board configuration(bus ID, chip select ID, mode, rates) in SPIDeviceManager::_device[] in libraries/AP_HAL_Linux/SPIDevice.cpp.
+
+
+GPIO Support
+------------
+Most ArduPilot Linux boards use one of two methods for GPIO access: sysfs or mmap.
+
+For sysfs:
+
+- Create new libraries/AP_HAL_Linux/GPIO_$(NEW_BOARD_NAME).cpp and .h, filled out with the needed _$(NEW_BOARD_NAME)_GPIO_MAX and pin_table.
+
+For mmap():
+
+- See GPIO_BBB.cpp and GPIO_RPI.cpp as examples
+- Create new libraries/AP_HAL_Linux/GPIO_$(NEW_BOARD_NAME).cpp and .h
+- Locate the memory file, either /dev/mem or /dev/gpiomem, and open() it.
+- Locate the offset of the GPIO memory region in the file. If using /dev/mem, check the processor reference manual or Linux device-tree files. For /dev/gpiomem, use zero.
+- mmap() the GPIO registers into ArduPilot using the above offset.
+- Implement pinMode(), read(), write() functions using the memory mapped GPIO registers.
+
+.. tip:: CPUs with multiple GPIOs per bank with only a DATA or OUT register without SET and CLEAR registers may need to use sysfs, as read-modify-write accesses to a GPIO DATA register may not be threadsafe when the kernel or other userspace programs use GPIOs in the same bank.
+
+For either approach, edit libraries/AP_HAL_Linux/GPIO.h to include the appropriate GPIO_ header for the board.
+
+RCInput
+-------
+Several different RCInput mechanisms are supported by different boards, including PPM, SBUS, UART, and UDP.
+
+PPM input is handled per-CPU, usually with either a DMA or utility microcontroller like the PRU on TI chips.
+
+Examine the libraries/AP_HAL_Linux/RCInput_* files to see if a suitable one exists, or create one.
+
+Configure rcinDriver in libraries/AP_HAL_Linux/HAL_Linux_Class.cpp using CONFIG_HAL_BOARD_SUBTYPE.
+
+RCOutput
+--------
+Similar to RCInput, examine the libraries/AP_HAL_Linux/RCOutput_* files to see if a suitable one exists, or create one.
+
+CPUs without corresponding hardware to generate RCOutput signals, like NXP's PCA9685 used for PWM generation.
+
+Configure the rcoutDriver in libraries/AP_HAL_Linux/HAL_Linux_Class.cpp using CONFIG_HAL_BOARD_SUBTYPE.
+
+Step 7 - bring up the sensors
+=============================
+
+Add sensor related configuration to libraries/AP_HAL/board/linux.h.
+
+Start with the baro first, then IMU, then compass and finally any other sensors
+the default sensor orientation should also be filled in along with other things.
```

Comment:
what I found really usefull is to use ardupilot examples. Those allow to each most part of ArduPilot libraries without having the mess of the full autopilot. Really usefull to test sensor implentation